File: vibelign/core/secret_scan.py
# === ANCHOR: SECRET_SCAN_START ===
from __future__ import annotations

import re
import shutil
import subprocess
import sys
from collections.abc import Callable, Iterable, Iterator
from dataclasses import dataclass
from pathlib import Path

from vibelign.core.structure_policy import WINDOWS_SUBPROCESS_FLAGS

def _find_git() -> str:
    found = shutil.which("git")
    if found:
        return found
    if sys.platform == "win32":
        # cmd\git.exe 는 CMD 래퍼라서 PyInstaller 환경에서 exit 129 발생 가능.
        # mingw64\bin\git.exe (실제 바이너리)를 먼저 시도.
        candidates = [
            r"C:\Program Files\Git\mingw64\bin\git.exe",
            r"C:\Program Files\Git\usr\bin\git.exe",
            r"C:\Program Files\Git\bin\git.exe",
            r"C:\Program Files\Git\cmd\git.exe",
            r"C:\Program Files (x86)\Git\mingw64\bin\git.exe",
            r"C:\Program Files (x86)\Git\cmd\git.exe",
            r"C:\Program Files (Arm)\Git\mingw64\bin\git.exe",
            r"C:\Program Files (Arm)\Git\cmd\git.exe",
        ]
        for p in candidates:
            if Path(p).exists():
                return p
    raise FileNotFoundError("git 실행 파일을 찾을 수 없어요. Git을 설치하고 PATH에 추가해주세요.")


_ALLOW_MARKER = "vibelign: allow-secret"
_PLACEHOLDER_VALUES = {
    "ENV",
    "YOUR_API_KEY",
    "YOUR_KEY",
    "YOUR_TOKEN",
    "YOUR_SECRET",
    "EXAMPLE",
    "CHANGE_ME",
    "CHANGEME",
    "CHANGE-ME",
    "REPLACE_ME",
    "REPLACE_WITH_REAL_VALUE",
}
_KEYWORD_VALUE_RE = re.compile(
    r"(?i)(api[_-]?key|token|secret|password|passwd|client_secret|access_key)\s*[:=]\s*[\"']?([A-Za-z0-9_./+=:@\-]{16,})[\"']?"
)
_HIGH_CONFIDENCE_RULES = [
    ("private-key", re.compile(r"-----BEGIN [A-Z ]*PRIVATE KEY-----")),
    (
        "github-token",
        re.compile(r"\b(?:ghp_[A-Za-z0-9]{20,}|github_pat_[A-Za-z0-9_]{20,})\b"),
    ),
    ("slack-token", re.compile(r"\bxox(?:a|b|p|r|s)-[A-Za-z0-9-]{10,}\b")),
    ("stripe-live-key", re.compile(r"\bsk_live_[A-Za-z0-9]{16,}\b")),
    ("aws-access-key", re.compile(r"\b(?:AKIA|ASIA)[0-9A-Z]{16}\b")),
    ("gemini-api-key", re.compile(r"\bAIzaSy[A-Za-z0-9_-]{33}\b")),
    ("anthropic-api-key", re.compile(r"\bsk-ant-[A-Za-z0-9_-]{40,}\b")),
    ("openai-api-key", re.compile(r"\bsk-[A-Za-z0-9]{32,}\b")),
    ("url-inline-key", re.compile(r"[?&]key=[A-Za-z0-9_-]{16,}")),
    (
        "jwt-token",
        re.compile(
            r"\beyJ[A-Za-z0-9_-]{8,}\.eyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\b"
        ),
    ),
    (
        "db-url-with-password",
        re.compile(
            r"\b(?:postgres(?:ql)?|mysql|mongodb(?:\+srv)?|redis|amqp|amqps)://"
            r"[^:/\s\"']+:[^@\s\"']+@[^\s\"']+"
        ),
    ),
    (
        "gcp-service-account",
        re.compile(r'"type"\s*:\s*"service_account"'),
    ),
]
_BINARY_SECRET_PATHS = {
    ".env",
    ".env.local",
    ".env.production",
    ".env.development",
    ".aws/credentials",
    "id_rsa",
}
_BINARY_SECRET_SUFFIXES = (".pem", ".key", ".p12", ".pfx")


# === ANCHOR: SECRET_SCAN_SECRETFINDING_START ===
@dataclass(frozen=True)
class SecretFinding:
    path: str
    rule_id: str
    line_number: int | None
    snippet: str
# === ANCHOR: SECRET_SCAN_SECRETFINDING_END ===


# === ANCHOR: SECRET_SCAN_SECRETSCANRESULT_START ===
@dataclass(frozen=True)
class SecretScanResult:
    findings: list[SecretFinding]

    # === ANCHOR: SECRET_SCAN_HAS_FINDINGS_START ===
    @property
    def has_findings(self) -> bool:
        return bool(self.findings)
    # === ANCHOR: SECRET_SCAN_HAS_FINDINGS_END ===
# === ANCHOR: SECRET_SCAN_SECRETSCANRESULT_END ===


# === ANCHOR: SECRET_SCAN__RUN_GIT_START ===
def _run_git(root: Path, args: list[str]) -> str:
    try:
        completed = subprocess.run(
            [_find_git(), *args],
            cwd=root,
            check=True,
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            creationflags=WINDOWS_SUBPROCESS_FLAGS,
        )
        return completed.stdout
    except subprocess.CalledProcessError as e:
        stderr = (e.stderr or "").strip()
        combined = (stderr + " " + (e.stdout or "")).lower()
        if "not a git repository" in combined or ("unknown option" in combined and "cached" in combined):
            raise RuntimeError("git 저장소가 아닌 폴더예요. 먼저 git init 을 실행해 주세요.") from None
        raise RuntimeError(
            f"git 명령 실패 (exit {e.returncode})"
            + (f": {stderr}" if stderr else "")
        ) from None
# === ANCHOR: SECRET_SCAN__RUN_GIT_END ===


# === ANCHOR: SECRET_SCAN__LOOKS_LIKE_SECRET_FILE_START ===
def _looks_like_secret_file(path: str) -> bool:
    name = path.replace("\\", "/")
    basename = Path(name).name
    return (
        name in _BINARY_SECRET_PATHS
        or basename in _BINARY_SECRET_PATHS
        or name.endswith(_BINARY_SECRET_SUFFIXES)
        or basename.endswith(_BINARY_SECRET_SUFFIXES)
    )
# === ANCHOR: SECRET_SCAN__LOOKS_LIKE_SECRET_FILE_END ===


# === ANCHOR: SECRET_SCAN__IS_PLACEHOLDER_START ===
def _is_placeholder(value: str) -> bool:
    normalized = value.strip().strip("\"'").upper()
    return normalized in _PLACEHOLDER_VALUES or normalized.startswith("YOUR_")
# === ANCHOR: SECRET_SCAN__IS_PLACEHOLDER_END ===


# === ANCHOR: SECRET_SCAN__REDACT_START ===
def _redact(text: str) -> str:
    compact = " ".join(text.strip().split())
    if len(compact) <= 4:
        return "[redacted]"
    return f"...{compact[-4:]}"
# === ANCHOR: SECRET_SCAN__REDACT_END ===


# === ANCHOR: SECRET_SCAN__EXTRACT_ADDED_LINE_START ===
def _extract_added_line(line: str) -> str | None:
    if not line.startswith("+") or line.startswith("+++"):
        return None
    return line[1:]
# === ANCHOR: SECRET_SCAN__EXTRACT_ADDED_LINE_END ===


# === ANCHOR: SECRET_SCAN_SCAN_UNIFIED_DIFF_FOR_SECRETS_START ===
def scan_unified_diff_for_secrets(
    diff_text: str, path_hint: str
# === ANCHOR: SECRET_SCAN_SCAN_UNIFIED_DIFF_FOR_SECRETS_END ===
) -> list[SecretFinding]:
    findings: list[SecretFinding] = []
    current_line_number: int | None = None

    for raw_line in diff_text.splitlines():
        if raw_line.startswith("@@"):
            match = re.search(r"\+(\d+)", raw_line)
            current_line_number = int(match.group(1)) if match else None
            continue

        added_line = _extract_added_line(raw_line)
        if added_line is None:
            if raw_line.startswith("-") or raw_line.startswith("diff --git"):
                continue
            if current_line_number is not None and not raw_line.startswith("\\"):
                current_line_number += 1
            continue

        line_number = current_line_number
        if current_line_number is not None:
            current_line_number += 1

        if _ALLOW_MARKER in added_line:
            continue

        matched_high_confidence = False
        for rule_id, pattern in _HIGH_CONFIDENCE_RULES:
            matched = pattern.search(added_line)
            if matched:
                matched_high_confidence = True
                findings.append(
                    SecretFinding(
                        path=path_hint,
                        rule_id=rule_id,
                        line_number=line_number,
                        snippet=_redact(matched.group(0)),
                    )
                )

        keyword_match = _KEYWORD_VALUE_RE.search(added_line)
        if not keyword_match or matched_high_confidence:
            continue

        value = keyword_match.group(2)
        if _is_placeholder(value):
            continue
        findings.append(
            SecretFinding(
                path=path_hint,
                rule_id="generic-secret",
                line_number=line_number,
                snippet=_redact(value),
            )
        )

    return findings


# === ANCHOR: SECRET_SCAN_SCAN_STAGED_SECRETS_START ===
def scan_staged_secrets(root: Path) -> SecretScanResult:
    if not (root / ".git").is_dir():
        return SecretScanResult(findings=[])
    findings: list[SecretFinding] = []
    names_output = _run_git(root, ["diff", "--cached", "--name-only", "-z"])
    staged_paths = [item for item in names_output.split("\0") if item]

    for path in staged_paths:
        if _looks_like_secret_file(path):
            findings.append(
                SecretFinding(
                    path=path,
                    rule_id="secret-file",
                    line_number=None,
                    snippet="secret-like file path",
                )
            )
            continue
        diff_text = _run_git(root, ["diff", "--cached", "--unified=0", "--", path])
        if diff_text.startswith("Binary files") or "GIT binary patch" in diff_text:
            continue
        findings.extend(scan_unified_diff_for_secrets(diff_text, path))

    return SecretScanResult(findings=findings)
# === ANCHOR: SECRET_SCAN_SCAN_STAGED_SECRETS_END ===


_COMMIT_MARKER_PREFIX = "COMMIT_MARKER_"
_DIFF_GIT_RE = re.compile(r"^diff --git a/(.*?) b/")

_HISTORY_AUDIT_SKIP_PATHS: tuple[str, ...] = (
    "tests/test_secret_scan.py",
)


def _is_history_audit_skipped(path: str) -> bool:
    return path in _HISTORY_AUDIT_SKIP_PATHS


# === ANCHOR: SECRET_SCAN_PARSE_GIT_LOG_CHUNKS_START ===
def parse_git_log_chunks(
    lines: Iterable[str],
) -> Iterator[tuple[str, str, str]]:
    current_commit = ""
    current_file = ""
    buffer: list[str] = []

    def _flush() -> tuple[str, str, str] | None:
        nonlocal buffer, current_file
        if buffer and current_file:
            out = (current_commit, current_file, "\n".join(buffer))
            buffer = []
            current_file = ""
            return out
        buffer = []
        current_file = ""
        return None

    for raw in lines:
        line = raw.rstrip("\r\n")
        if line.startswith(_COMMIT_MARKER_PREFIX):
            flushed = _flush()
            if flushed is not None:
                yield flushed
            current_commit = line[len(_COMMIT_MARKER_PREFIX):]
            continue
        m = _DIFF_GIT_RE.match(line)
        if m is not None:
            flushed = _flush()
            if flushed is not None:
                yield flushed
            current_file = m.group(1)
            buffer.append(line)
            continue
        if current_file:
            buffer.append(line)

    tail = _flush()
    if tail is not None:
        yield tail
# === ANCHOR: SECRET_SCAN_PARSE_GIT_LOG_CHUNKS_END ===


# === ANCHOR: SECRET_SCAN_SCAN_ALL_HISTORY_START ===
def scan_all_history(
    root: Path,
    on_progress: Callable[[int, int | None], None] | None = None,
) -> SecretScanResult:
    if not (root / ".git").is_dir():
        return SecretScanResult(findings=[])

    total: int | None = None
    try:
        total_out = _run_git(root, ["rev-list", "--count", "--all"])
        total = int(total_out.strip())
    except (RuntimeError, ValueError):
        total = None

    findings: list[SecretFinding] = []

    try:
        git_bin = _find_git()
    except FileNotFoundError:
        return SecretScanResult(findings=[])

    proc = subprocess.Popen(
        [
            git_bin,
            "log",
            "--all",
            "-p",
            "--no-color",
            f"--format={_COMMIT_MARKER_PREFIX}%H",
        ],
        cwd=root,
        stdout=subprocess.PIPE,
        stderr=subprocess.DEVNULL,
        encoding="utf-8",
        errors="replace",
        creationflags=WINDOWS_SUBPROCESS_FLAGS,
    )
    assert proc.stdout is not None

    try:
        last_reported = None
        processed = 0
        for commit_sha, file_path, diff_text in parse_git_log_chunks(proc.stdout):
            display_path = (
                f"{commit_sha[:8]}:{file_path}" if commit_sha else file_path
            )
            if _is_history_audit_skipped(file_path):
                if commit_sha != last_reported:
                    last_reported = commit_sha
                    processed += 1
                    if on_progress is not None:
                        on_progress(processed, total)
                continue
            if _looks_like_secret_file(file_path):
                findings.append(
                    SecretFinding(
                        path=display_path,
                        rule_id="secret-file",
                        line_number=None,
                        snippet="secret-like file path",
                    )
                )
            elif not (
                diff_text.startswith("Binary files") or "GIT binary patch" in diff_text
            ):
                findings.extend(
                    scan_unified_diff_for_secrets(diff_text, display_path)
                )

            if commit_sha != last_reported:
                last_reported = commit_sha
                processed += 1
                if on_progress is not None:
                    on_progress(processed, total)
    finally:
        try:
            proc.stdout.close()
        except OSError:
            pass
        try:
            _ = proc.wait(timeout=5)
        except subprocess.TimeoutExpired:
            proc.kill()
            _ = proc.wait()

    return SecretScanResult(findings=findings)
# === ANCHOR: SECRET_SCAN_SCAN_ALL_HISTORY_END ===
# === ANCHOR: SECRET_SCAN_END ===

File: vibelign-core/src/secret_scan.rs
// === ANCHOR: SECRET_SCAN_START ===
//! Pure-function port of Python `vibelign.core.secret_scan.scan_unified_diff_for_secrets`.
//!
//! Python is the source of truth — see `tests/fixtures/secret_scan_diffs/*.expected.json`.
//! Git invocation and file walking stay in Python; this module only does regex matching
//! on a single (diff_text, path_hint) pair, which is the CPU-hot work.

use regex::Regex;
use serde::Serialize;
use std::sync::OnceLock;

#[derive(Debug, Clone, Serialize, PartialEq, Eq)]
pub struct SecretFinding {
    pub path: String,
    pub rule_id: String,
    pub line_number: Option<u32>,
    pub snippet: String,
}

const ALLOW_MARKER: &str = "vibelign: allow-secret";

const PLACEHOLDER_VALUES: &[&str] = &[
    "ENV",
    "YOUR_API_KEY",
    "YOUR_KEY",
    "YOUR_TOKEN",
    "YOUR_SECRET",
    "EXAMPLE",
    "CHANGE_ME",
    "CHANGEME",
    "CHANGE-ME",
    "REPLACE_ME",
    "REPLACE_WITH_REAL_VALUE",
];

struct HighConfidenceRule {
    rule_id: &'static str,
    pattern: &'static str,
}

const HIGH_CONFIDENCE_RULES: &[HighConfidenceRule] = &[
    HighConfidenceRule { rule_id: "private-key", pattern: r"-----BEGIN [A-Z ]*PRIVATE KEY-----" },
    HighConfidenceRule {
        rule_id: "github-token",
        pattern: r"\b(?:ghp_[A-Za-z0-9]{20,}|github_pat_[A-Za-z0-9_]{20,})\b",
    },
    HighConfidenceRule { rule_id: "slack-token", pattern: r"\bxox(?:a|b|p|r|s)-[A-Za-z0-9-]{10,}\b" },
    HighConfidenceRule { rule_id: "stripe-live-key", pattern: r"\bsk_live_[A-Za-z0-9]{16,}\b" },
    HighConfidenceRule { rule_id: "aws-access-key", pattern: r"\b(?:AKIA|ASIA)[0-9A-Z]{16}\b" },
    HighConfidenceRule { rule_id: "gemini-api-key", pattern: r"\bAIzaSy[A-Za-z0-9_-]{33}\b" },
    HighConfidenceRule { rule_id: "anthropic-api-key", pattern: r"\bsk-ant-[A-Za-z0-9_-]{40,}\b" },
    HighConfidenceRule { rule_id: "openai-api-key", pattern: r"\bsk-[A-Za-z0-9]{32,}\b" },
    HighConfidenceRule { rule_id: "url-inline-key", pattern: r"[?&]key=[A-Za-z0-9_-]{16,}" },
    HighConfidenceRule {
        rule_id: "jwt-token",
        pattern: r"\beyJ[A-Za-z0-9_-]{8,}\.eyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\b",
    },
    HighConfidenceRule {
        rule_id: "db-url-with-password",
        pattern: r"\b(?:postgres(?:ql)?|mysql|mongodb(?:\+srv)?|redis|amqp|amqps)://[^:/\s\x22']+:[^@\s\x22']+@[^\s\x22']+",
    },
    HighConfidenceRule { rule_id: "gcp-service-account", pattern: r#""type"\s*:\s*"service_account""# },
];

const KEYWORD_VALUE_PATTERN: &str =
    r"(?i)(api[_-]?key|token|secret|password|passwd|client_secret|access_key)\s*[:=]\s*[\x22']?([A-Za-z0-9_./+=:@\-]{16,})[\x22']?";

const HUNK_HEADER_PLUS_PATTERN: &str = r"\+(\d+)";

fn high_confidence_regexes() -> &'static [(&'static str, Regex)] {
    static CACHE: OnceLock<Vec<(&'static str, Regex)>> = OnceLock::new();
    CACHE.get_or_init(|| {
        HIGH_CONFIDENCE_RULES
            .iter()
            .map(|rule| (rule.rule_id, Regex::new(rule.pattern).expect("static high-confidence rule must compile")))
            .collect()
    })
}

fn keyword_regex() -> &'static Regex {
    static CACHE: OnceLock<Regex> = OnceLock::new();
    CACHE.get_or_init(|| Regex::new(KEYWORD_VALUE_PATTERN).expect("static keyword pattern must compile"))
}

fn hunk_plus_regex() -> &'static Regex {
    static CACHE: OnceLock<Regex> = OnceLock::new();
    CACHE.get_or_init(|| Regex::new(HUNK_HEADER_PLUS_PATTERN).expect("static hunk header pattern must compile"))
}

fn is_placeholder(value: &str) -> bool {
    let trimmed = value.trim().trim_matches(|character| character == '"' || character == '\'');
    let normalized = trimmed.to_uppercase();
    if PLACEHOLDER_VALUES.iter().any(|placeholder| placeholder == &normalized.as_str()) {
        return true;
    }
    normalized.starts_with("YOUR_")
}

fn redact(text: &str) -> String {
    let compact: String = text.split_whitespace().collect::<Vec<_>>().join(" ");
    if compact.chars().count() <= 4 {
        return "[redacted]".to_string();
    }
    let tail: String = compact.chars().rev().take(4).collect::<String>().chars().rev().collect();
    format!("...{tail}")
}

fn extract_added_line(line: &str) -> Option<&str> {
    if !line.starts_with('+') || line.starts_with("+++") {
        return None;
    }
    Some(&line[1..])
}

pub fn scan_unified_diff(diff_text: &str, path_hint: &str) -> Vec<SecretFinding> {
    let mut findings: Vec<SecretFinding> = Vec::new();
    let mut current_line_number: Option<u32> = None;

    for raw_line in diff_text.split('\n') {
        let line = raw_line.strip_suffix('\r').unwrap_or(raw_line);

        if line.starts_with("@@") {
            current_line_number = hunk_plus_regex()
                .captures(line)
                .and_then(|captures| captures.get(1))
                .and_then(|matched| matched.as_str().parse::<u32>().ok());
            continue;
        }

        let added_line = match extract_added_line(line) {
            Some(content) => content,
            None => {
                if line.starts_with('-') || line.starts_with("diff --git") {
                    continue;
                }
                if let Some(counter) = current_line_number.as_mut() {
                    if !line.starts_with('\\') {
                        *counter = counter.saturating_add(1);
                    }
                }
                continue;
            }
        };

        let line_number_for_finding = current_line_number;
        if let Some(counter) = current_line_number.as_mut() {
            *counter = counter.saturating_add(1);
        }

        if added_line.contains(ALLOW_MARKER) {
            continue;
        }

        let mut matched_high_confidence = false;
        for (rule_id, pattern) in high_confidence_regexes() {
            if let Some(matched) = pattern.find(added_line) {
                matched_high_confidence = true;
                findings.push(SecretFinding {
                    path: path_hint.to_string(),
                    rule_id: (*rule_id).to_string(),
                    line_number: line_number_for_finding,
                    snippet: redact(matched.as_str()),
                });
            }
        }

        if matched_high_confidence {
            continue;
        }

        let Some(keyword_match) = keyword_regex().captures(added_line) else {
            continue;
        };
        let Some(value_group) = keyword_match.get(2) else {
            continue;
        };
        let value = value_group.as_str();
        if is_placeholder(value) {
            continue;
        }
        findings.push(SecretFinding {
            path: path_hint.to_string(),
            rule_id: "generic-secret".to_string(),
            line_number: line_number_for_finding,
            snippet: redact(value),
        });
    }

    findings
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;

    fn fixtures_dir() -> PathBuf {
        let manifest = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        manifest.parent().expect("workspace root").join("tests/fixtures/secret_scan_diffs")
    }

    fn path_hint_from_golden(diff_name: &str) -> String {
        let dir = fixtures_dir();
        let expected_name = format!("{}.expected.json", diff_name.trim_end_matches(".diff"));
        let expected_text = std::fs::read_to_string(dir.join(&expected_name)).expect("fixture .expected.json exists");
        let expected: serde_json::Value = serde_json::from_str(&expected_text).expect("valid expected json");
        expected["path_hint"].as_str().expect("path_hint string").to_string()
    }

    fn run_fixture(diff_name: &str) -> Vec<SecretFinding> {
        let dir = fixtures_dir();
        let diff_text = std::fs::read_to_string(dir.join(diff_name)).expect("fixture .diff exists");
        scan_unified_diff(&diff_text, &path_hint_from_golden(diff_name))
    }

    fn expected_findings(diff_name: &str) -> Vec<SecretFinding> {
        let dir = fixtures_dir();
        let expected_name = format!("{}.expected.json", diff_name.trim_end_matches(".diff"));
        let expected_text = std::fs::read_to_string(dir.join(&expected_name)).expect("fixture .expected.json exists");
        let expected: serde_json::Value = serde_json::from_str(&expected_text).expect("valid expected json");
        let findings = expected["findings"].as_array().expect("findings array").clone();
        findings
            .into_iter()
            .map(|item| SecretFinding {
                path: item["path"].as_str().unwrap_or_default().to_string(),
                rule_id: item["rule_id"].as_str().unwrap_or_default().to_string(),
                line_number: item["line_number"].as_u64().map(|number| number as u32),
                snippet: item["snippet"].as_str().unwrap_or_default().to_string(),
            })
            .collect()
    }

    fn assert_parity(diff_name: &str) {
        let actual = run_fixture(diff_name);
        let expected = expected_findings(diff_name);
        assert_eq!(actual, expected, "fixture {} parity mismatch", diff_name);
    }

    #[test]
    fn empty_diff_yields_no_findings() {
        assert_parity("01_empty.diff");
    }

    #[test]
    fn high_confidence_aws_rule_fires() {
        assert_parity("02_high_confidence_aws.diff");
    }

    #[test]
    fn high_confidence_combo_emits_one_finding_per_rule() {
        assert_parity("03_high_confidence_combo.diff");
    }

    #[test]
    fn keyword_rule_skips_placeholder() {
        assert_parity("04_keyword_with_placeholder.diff");
    }

    #[test]
    fn allow_marker_suppresses_findings_on_that_line() {
        assert_parity("05_allow_marker.diff");
    }

    #[test]
    fn line_numbers_track_across_hunks() {
        assert_parity("06_line_number_tracking.diff");
    }

    #[test]
    fn removed_line_does_not_emit_finding() {
        assert_parity("07_removed_line_ignored.diff");
    }

    #[test]
    fn no_newline_marker_does_not_increment_line_counter() {
        assert_parity("08_no_newline_marker.diff");
    }

    #[test]
    fn high_confidence_wins_over_keyword_on_same_line() {
        assert_parity("09_high_confidence_wins.diff");
    }

    #[test]
    fn unicode_path_and_body_round_trip_intact() {
        assert_parity("10_unicode_path_and_body.diff");
    }
}
// === ANCHOR: SECRET_SCAN_END ===
